### ISSUE TITLE
Make `timestamp_tuple` public to enable mocking

### ISSUE THREAD

**Issue by jesenko** [elixir-lang/ecto#1007]
Issue is related to this [question](http://stackoverflow.com/questions/32903786/time-travel-and-time-freezing-in-elixir). In summary, I would like to mock current time in integration tests; It seems that it is impossible (or highly untrivial) to mock `:os.current_time` in erlang/elixir, which is used in `Ecto.Model.Timestamps` to provide `inserted_at`/`updated_at` values.

Currently, I could mock `put_timestamp` function using `meck` to provide mocked current time, however I would basically have to replicate all the logic that is currently in `Ecto.Model.Timestamps` module.

If `timestamp_tuple\1` would be public, it would be much easier to mock current time in Ecto.

**Comment by josevalim** [comment]
We are definitely not making it public, specially if it is for mocking
purposes. It is a really bad sign if you want to mock private function. It
is an implementation detail, you are not supposed to even know about it
without the source code. Can you instead tell us what you are trying to do?

-- 


*José Valim*
www.plataformatec.com.br
Skype: jv.ptec
Founder and Director of R&D

**Comment by jesenko** [comment]
I am trying to write integration tests for some functionality that is directly related to the current system time, e.g. filtering events that are received within 1 minute of each other. Maybe best example will be using some hypothetical code:

```elixir
# within e.g. phoenix controller
def add_event(conn, params) do
  # Using custom provider for current time which enables 
  # explicit setting of current time in integration tests
  current_time = MyTimeProvider.current_time
  offset_time = add_minute(current_time, -1)
  should_filter =  
    Repo.all(from(e in Event, where: inserted_at > offset_time , select: count(e.id))) > 0
  unless should_filter do
    # yeah, I could explicitly set inserted_at to my custom `current_time`,
    # however, to be consistent, I should probably do it _everywhere_, ditching 
    # Ecto.Model.Timestamp support altogether 
    Repo.insert!(Event.cast(params))
  end
end

# within integration test
test "filtering of events" do
  set_current_time(2010, 1, 1, 0, 0, 0, 0)
  post_event(%{...})
  set_current_time(2010, 1, 1, 0, 0, 0, 30)
  post_event(%{...})
  assert_event_not_added
  set_current_time(2010, 1, 1, 0, 0, 1, 30)
  post_event(%{...})
  assert_event_added
end
```

I am using `MyTimeProvider` everywhere in my code instead of `:os.timestamp`, `:calendar.universal_time` etc, so that I can easily specify it in integration test. I cannot however do much when external deps use them directly as they are essentially [_unmockable_](http://stackoverflow.com/questions/32903786/time-travel-and-time-freezing-in-elixir).

As Ecto does not allow specifying `alternative` time provider, currently my only option would be to ditch `Ecto.Model.Timestamp` altogether, and set `inserted_at` and `updated_at` explicitly whenever I am inserting / updating records.

After thinking about it, I must agree that making `timestamp_tuple` public just for sake of testing would be indeed _very_ ugly.

What about enabling explicitly setting function that provides timestamp in configuration (same as `inserted_at` field can be set)? E.g.:
```elixir
  def model do
    quote do
      use Ecto.Model
      @timestamps_opts [timestamp_tuple: MyTimeProvider.timestamp_tuple]
    end
  end
```

**Comment by josevalim** [comment]
I am still torn because you can easily define your own callbacks and your own changesets that will do that for you. For example, Event.cast could create a changeset where you put the timestamp value that you want.

**Comment by jesenko** [comment]
I guess the best solution for my case might be to implement custom module `MyEctoTimestamps` and then cherry-pick modules from `Ecto.Module` explicitly,

```elixir
def model do
  quote do
    use Ecto.Schema
    import Ecto.Changeset
    import Ecto.Query, only: [from: 2]

    import Ecto.Model
    use Ecto.Model.OptimisticLock
    use MyEctoTimestamps # use this instead of ecto timestamps
    use Ecto.Model.Dependent
    use Ecto.Model.Autogenerate
    use Ecto.Model.Callbacks
  end
end
```
My module `MyEctoTimestamps` will be more-or-less just a rip-off of `Ecto.Model.Timestamps` with custom `timestamp_tuple` method. With this solution, _all_ of my models would consistently use mocked time in tests, requiring no special handling of timestamps.

I guess it would still be nicer if I could just set `timestamp_tuple` function in `@timestamp_opts`. This option however might clash somewhat with current Timestamp module API, e.g. `usec` option in `@timestamp_opts` does not make sense when explicitly setting `timestamp_tuple` function...

In summary, I found a solution that works for me in current Ecto. Configuring `timestamp_tuple` might be worth reconsidering if this scenario comes up more often by other users. 

@josevalim Thank you for looking into it!

**Comment by josevalim** [comment]
You don't need to cherry pick the functionality necessarily as you can set this stuff when you set up the changeset or just run it on top of the Ecto one. :)

**Comment by jesenko** [comment]
@josevalim not sure I understand your last comment...
 > you can set this stuff when you set up the changeset

Do you mean explicitly setting up timestamps in each changeset before persisting it to db? I would very much like to avoid that, leaving it to Ecto to handle it. It might get very cumbersome if changeset has a nested association that also have timestamps, etc.
> or just run it on top of the Ecto one

You mean something like this?
```elixir
def model do
  quote do
    use Ecto.Model
    use MyEctoTimestamps # use this instead of ecto timestamps
  end
end
```
Hm I guess that in this case, `timestamp` macro would register callbacks to `Ecto.Model.Timestamps` as well as to `MyEctoTimestamps` - probably resulting in first-win situation, i.e. `Ecto.Model.Timestamps` setting timestamps.

Have I missed anything?

**Comment by josevalim** [comment]
>  It might get very cumbersome if changeset has a nested association that also have timestamps, etc.

Good point.

> Hm I guess that in this case, timestamp macro would register callbacks to Ecto.Model.Timestamps as well as to MyEctoTimestamps - probably resulting in first-win situation, i.e. Ecto.Model.Timestamps setting timestamps.

You should be fine. If you run *before*, Ecto timestamps will respect your value. If you run *after*, you can use force_change to override the Ecto one.

**Comment by jesenko** [comment]
> You should be fine. If you run before, Ecto timestamps will respect your value. If you run after, you can use force_change to override the Ecto one.

Great, will try this. Anyhow, even cherry-picking explicitly is not much of an issue, as I only have to do it once in app for all models. Thanks for suggestions!
